feat: Added read_time as a parameter to various calls (synchronous/base classes)

gkevinzheng opened the pull request:
This PR adds `read_time` to API calls that map to various Firestore gRPC calls whose request objects contain `read_time` as a field.

Calls with `read_time` added:
| Client Library Call                        | gRPC Request in Call  |
| --------------------------------| --------------- |
| `AggregationQuery.get/stream`| [RunAggregationQueryRequest](https://firebase.google.com/docs/firestore/reference/rpc/google.firestore.v1#runaggregationqueryrequest) |
| `Client.get_all` | [BatchGetDocumentsRequest](https://firebase.google.com/docs/firestore/reference/rpc/google.firestore.v1#batchgetdocumentsrequest) |
| `Client.collections` | [ListCollectionIdsRequest](https://firebase.google.com/docs/firestore/reference/rpc/google.firestore.v1#listcollectionidsrequest) |
| `Collection.list_documents` | [ListDocumentsRequest](https://firebase.google.com/docs/firestore/reference/rpc/google.firestore.v1#listdocumentsrequest) |
| `Collection.get/stream` | [RunQueryRequest](https://firebase.google.com/docs/firestore/reference/rpc/google.firestore.v1#runqueryrequest) |
| `DocumentReference.get` | [BatchGetDocumentsRequest](https://firebase.google.com/docs/firestore/reference/rpc/google.firestore.v1#batchgetdocumentsrequest) |
| `DocumentReference.collections` | [ListCollectionIdsRequest](https://firebase.google.com/docs/firestore/reference/rpc/google.firestore.v1#listcollectionidsrequest) |
| `Query.get/stream` | [RunQueryRequest](https://firebase.google.com/docs/firestore/reference/rpc/google.firestore.v1#runqueryrequest) |
| `Query.list_partitions` | [PartitionQueryRequest](https://firebase.google.com/docs/firestore/reference/rpc/google.firestore.v1#partitionqueryrequest) |
| `Transaction.get_all` | [BatchGetDocumentsRequest](https://firebase.google.com/docs/firestore/reference/rpc/google.firestore.v1#batchgetdocumentsrequest) |
| `Transaction.get` | [BatchGetDocumentsRequest](https://firebase.google.com/docs/firestore/reference/rpc/google.firestore.v1#batchgetdocumentsrequest)/ [RunQueryRequest](https://firebase.google.com/docs/firestore/reference/rpc/google.firestore.v1#runqueryrequest) |

Partially fixes #775 for synchronous/base classes, async classes will be in a separate PR.

daniel-sanche committed:
chore: update protoplus for python 3.13 (#1009)

gcf-owl-bot[bot] committed:
fix: bump default deadline on CreateDatabase and RestoreDatabase to 2 minutes (#975)

* chore: Update gapic-generator-python to v1.19.1

PiperOrigin-RevId: 684571179

Source-Link: https://github.com/googleapis/googleapis/commit/fbdc238931e0a7a95c0f55e0cd3ad9e3de2535c8

Source-Link: https://github.com/googleapis/googleapis-gen/commit/3a2cdcfb80c2d0f5ec0cc663c2bab0a9486229d0
Copy-Tag: eyJwIjoiLmdpdGh1Yi8uT3dsQm90LnlhbWwiLCJoIjoiM2EyY2RjZmI4MGMyZDBmNWVjMGNjNjYzYzJiYWIwYTk0ODYyMjlkMCJ9

* 🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

* chore: Configure Ruby clients for google-ads-ad_manager

PiperOrigin-RevId: 689139590

Source-Link: https://github.com/googleapis/googleapis/commit/296f2ac1aa9abccb7708b639b7839faa1809087f

Source-Link: https://github.com/googleapis/googleapis-gen/commit/26927362e0aa1293258fc23fe3ce83c5c21d5fbb
Copy-Tag: eyJwIjoiLmdpdGh1Yi8uT3dsQm90LnlhbWwiLCJoIjoiMjY5MjczNjJlMGFhMTI5MzI1OGZjMjNmZTNjZTgzYzVjMjFkNWZiYiJ9

* 🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

* chore: remove body selector from http rule

PiperOrigin-RevId: 693215877

Source-Link: https://github.com/googleapis/googleapis/commit/bb6b53e326ce2db403d18be7158c265e07948920

Source-Link: https://github.com/googleapis/googleapis-gen/commit/db8b5a93484ad44055b2bacc4c7cf87e970fe0ed
Copy-Tag: eyJwIjoiLmdpdGh1Yi8uT3dsQm90LnlhbWwiLCJoIjoiZGI4YjVhOTM0ODRhZDQ0MDU1YjJiYWNjNGM3Y2Y4N2U5NzBmZTBlZCJ9

* 🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

* feat: Add support for opt-in debug logging
fix: Fix typing issue with gRPC metadata when key ends in -bin
chore: Update gapic-generator-python to v1.21.0

PiperOrigin-RevId: 705285820

Source-Link: https://github.com/googleapis/googleapis/commit/f9b8b9150f7fcd600b0acaeef91236b1843f5e49

Source-Link: https://github.com/googleapis/googleapis-gen/commit/ca1e0a1e472d6e6f5de883a5cb54724f112ce348
Copy-Tag: eyJwIjoiLmdpdGh1Yi8uT3dsQm90LnlhbWwiLCJoIjoiY2ExZTBhMWU0NzJkNmU2ZjVkZTg4M2E1Y2I1NDcyNGYxMTJjZTM0OCJ9

* 🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

* fix: bump default deadline on CreateDatabase and RestoreDatabase to 2 minutes
feat: add filter argument to FirestoreAdmin.ListBackupsRequest

PiperOrigin-RevId: 716763143

Source-Link: https://github.com/googleapis/googleapis/commit/3776db131e34e42ec8d287203020cb4282166aa5

Source-Link: https://github.com/googleapis/googleapis-gen/commit/10db5ac476a94aa4c9e0a24946d9fa1b7ea456f6
Copy-Tag: eyJwIjoiLmdpdGh1Yi8uT3dsQm90LnlhbWwiLCJoIjoiMTBkYjVhYzQ3NmE5NGFhNGM5ZTBhMjQ5NDZkOWZhMWI3ZWE0NTZmNiJ9

* 🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

---------

Co-authored-by: Owl Bot <gcf-owl-bot[bot]@users.noreply.github.com>
Co-authored-by: Daniel Sanche <sanche@google.com>

gkevinzheng committed:
feat: Added read_time as a parameter to various calls (synchronous/base
classes)

gkevinzheng committed:
feat: Added read_time as a parameter to various calls (synchronous/base
classes)

gkevinzheng committed:
Added system test for DocumentReference.collections

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

gkevinzheng committed:
linting

gkevinzheng committed:
Conditionally adding read_time to kwargs for cleaner test separation
between sync/async

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

gkevinzheng committed:
Conditionally adding read_time to kwargs in base_aggregation.py

gkevinzheng committed:
More conditionally adding read_time to kwargs

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

gkevinzheng committed:
Added tests for collections.get/stream

gkevinzheng committed:
linting

daniel-sanche commented on google/cloud/firestore_v1/aggregation.py:
```diff
@@ -20,6 +20,7 @@
 """
 from __future__ import annotations
 
+from datetime import datetime
```
nit: just importing datetime as a module is probably better? This import overrides the "datetime" module name with the class name, which would get a bit weird if we have to use other datetime imports in the future

And it makes some of the type annotations a bit ambiguous

daniel-sanche commented on google/cloud/firestore_v1/aggregation.py:
```diff
@@ -78,6 +80,10 @@ def get(
+            read_time (Optional[datetime]): If set, reads documents as they were at the given
+                time. This must be a microsecond precision timestamp within the past one hour,
+                or if Point-in-Time Recovery is enabled, can additionally be a whole minute
+                timestamp within the past 7 days.
```
A couple questions:
- how are timezones handled? Do you expect ["aware" or "naive" datetimes?](https://docs.python.org/3/library/datetime.html#aware-and-naive-objects)
- I just did a quick search, but it seems like datetime.datetime also has microsecond precision. Do we need to mention the precision here?

daniel-sanche commented on google/cloud/firestore_v1/client.py:
```diff
@@ -201,10 +202,12 @@ def document(self, *document_path: str) -> DocumentReference:
-        field_paths: Iterable[str] | None = None,
-        transaction: Transaction | None = None,
+        field_paths: Iterable[str] = None,
+        transaction: Transaction = None,
```
why remove None from these? Doesn't that cause a mypy error?

gkevinzheng replied:
Whoops, my bad. This was most likely from merging a really old branch. I'll revert it.

daniel-sanche commented on google/cloud/firestore_v1/client.py:
```diff
@@ -201,10 +202,12 @@ def document(self, *document_path: str) -> DocumentReference:
+        transaction: Transaction = None,
         retry: retries.Retry | object | None = gapic_v1.method.DEFAULT,
-        timeout: float | None = None,
+        timeout: float = None,
```
same comment: isn't this a type violation?

daniel-sanche commented on google/cloud/firestore_v1/client.py:
```diff
@@ -239,13 +242,17 @@ def get_all(
+            read_time (Optional[datetime.datetime]): If set, reads documents as they were at the given
+                time. This must be a microsecond precision timestamp within the past one hour, or
+                if Point-in-Time Recovery is enabled, can additionally be a whole minute timestamp
+                within the past 7 days. For the most accurate results, use UTC timezone.
```
1. I'm confused by this comment. Why "most accurate results". Wouldn't the wrong timezone give wrong results? How are timezones meant to be used?
2. Shouldn't this docstring be consistent with others? I don't see UTC note this in the previous files

daniel-sanche commented on tests/system/test_system.py:
```diff
@@ -3206,6 +3426,52 @@ def in_transaction(transaction):
+        # make sure we didn't skip assertions in inner function
+        assert inner_fn_ran is True
+
+
```
it would be good to test a read_time that is far in the past, or in the future, to make sure the exception is raised as you expect it to be

daniel-sanche commented on tests/unit/v1/test_document.py:
```diff
@@ -520,7 +529,14 @@ def test_documentreference_get_with_transaction(database):
-def _collections_helper(page_size=None, retry=None, timeout=None, database=None):
+@pytest.mark.parametrize("database", [None, "somedb"])
+def test_documentreference_get_with_read_time(database):
+    _get_helper(read_time=DatetimeWithNanoseconds.now(), database=database)
```
Why use from DatetimeWithNanoseconds here instead of the regular datetime?

gkevinzheng replied:
The `_get_helper` creates a mock response which requires a `Timestamp` object for the read time, so I used `DatetimeWithNanoseconds` to get access to `read_time.timestamp_pb()` to get that field. I guess I could just use a regular `datetime` and have it be mocked accordingly.

gkevinzheng committed:
Addressed review feedback

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

gkevinzheng committed:
linting

daniel-sanche committed:
fix: client-side path validation for batch.update (#1021)

daniel-sanche committed:
fix: Watch thread deadlock on exit (#1014)

* fix: fix thread cleanup when destroying Watch instance

* added test for on_snapshot while closing

gcf-owl-bot[bot] committed:
chore(python): fix typo in README (#1015)

Source-Link: https://github.com/googleapis/synthtool/commit/93e1685311a3940e713fd00820aa9937d496f544
Post-Processor: gcr.io/cloud-devrel-public-resources/owlbot-python:latest@sha256:631b4a35a4f9dd5e97740a97c4c117646eb85b35e103844dc49d152bd18694cd

Co-authored-by: Owl Bot <gcf-owl-bot[bot]@users.noreply.github.com>
Co-authored-by: Daniel Sanche <sanche@google.com>

gcf-owl-bot[bot] committed:
chore(python): conditionally load credentials in .kokoro/build.sh (#1022)

Source-Link: https://github.com/googleapis/synthtool/commit/aa69fb74717c8f4c58c60f8cc101d3f4b2c07b09
Post-Processor: gcr.io/cloud-devrel-public-resources/owlbot-python:latest@sha256:f016446d6e520e5fb552c45b110cba3f217bffdd3d06bdddd076e9e6d13266cf

Co-authored-by: Owl Bot <gcf-owl-bot[bot]@users.noreply.github.com>
Co-authored-by: Daniel Sanche <sanche@google.com>

gcf-owl-bot[bot] committed:
chore: add logging section to readme (#1018)

* chore: add logging section to readme

Source-Link: https://github.com/googleapis/synthtool/commit/d1011bc72b89a605d8e72c1f45c1db119fd56d74
Post-Processor: gcr.io/cloud-devrel-public-resources/owlbot-python:latest@sha256:fdc038572b896f739f95cc90e62f16c06e4f2ef0ef3bea343a358331862ad0f0

* 🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

---------

Co-authored-by: Owl Bot <gcf-owl-bot[bot]@users.noreply.github.com>
Co-authored-by: Daniel Sanche <sanche@google.com>

gcf-owl-bot[bot] committed:
chore: pull up gapic updates (#1016)

* feat: Add REST Interceptors which support reading metadata
feat: Add support for reading selective GAPIC generation methods from service YAML
chore: Update gapic-generator-python to v1.22.0

PiperOrigin-RevId: 724026024

Source-Link: https://github.com/googleapis/googleapis/commit/ad9963857109513e77eed153a66264481789109f

Source-Link: https://github.com/googleapis/googleapis-gen/commit/e291c4dd1d670eda19998de76f967e1603a48993
Copy-Tag: eyJwIjoiLmdpdGh1Yi8uT3dsQm90LnlhbWwiLCJoIjoiZTI5MWM0ZGQxZDY3MGVkYTE5OTk4ZGU3NmY5NjdlMTYwM2E0ODk5MyJ9

* 🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

* chore: expose the Firestore.executePipeline API to the preview branch
docs: minor documentation updates to `StructuredQuery`
docs: minor documentation changes for `distance_threshold`

PiperOrigin-RevId: 731306872

Source-Link: https://github.com/googleapis/googleapis/commit/b6d5ae84d5070e249319dc898f81becf86546414

Source-Link: https://github.com/googleapis/googleapis-gen/commit/4cb9048871bf5906b390e8ea3d7b9f1ef6ac19d6
Copy-Tag: eyJwIjoiLmdpdGh1Yi8uT3dsQm90LnlhbWwiLCJoIjoiNGNiOTA0ODg3MWJmNTkwNmIzOTBlOGVhM2Q3YjlmMWVmNmFjMTlkNiJ9

* 🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

---------

Co-authored-by: Owl Bot <gcf-owl-bot[bot]@users.noreply.github.com>
Co-authored-by: Daniel Sanche <sanche@google.com>

release-please[bot] committed:
chore(main): release 2.20.1 (#1011)

Co-authored-by: release-please[bot] <55107282+release-please[bot]@users.noreply.github.com>
Co-authored-by: Daniel Sanche <sanche@google.com>

gcf-owl-bot[bot] committed:
chore: Update gapic-generator-python to v1.23.2 (#1024)

* chore: Update gapic-generator-python to v1.23.2

PiperOrigin-RevId: 732281673

Source-Link: https://github.com/googleapis/googleapis/commit/2f37e0ad56637325b24f8603284ccb6f05796f9a

Source-Link: https://github.com/googleapis/googleapis-gen/commit/016b7538ba5a798f2ae423d4ccd7f82b06cdf6d2
Copy-Tag: eyJwIjoiLmdpdGh1Yi8uT3dsQm90LnlhbWwiLCJoIjoiMDE2Yjc1MzhiYTVhNzk4ZjJhZTQyM2Q0Y2NkN2Y4MmIwNmNkZjZkMiJ9

* 🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

---------

Co-authored-by: Owl Bot <gcf-owl-bot[bot]@users.noreply.github.com>

parthea committed:
chore: remove unused files (#1027)

* chore: remove unused files

* 🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

---------

Co-authored-by: Owl Bot <gcf-owl-bot[bot]@users.noreply.github.com>

parthea committed:
fix: allow Protobuf 6.x (#1028)

gcf-owl-bot[bot] committed:
chore: Update gapic-generator-python to 1.23.6 (#1032)

* chore: Update gapic-generator-python to 1.23.6

PiperOrigin-RevId: 738170370

Source-Link: https://github.com/googleapis/googleapis/commit/3f1e17aa2dec3f146a9a2a8a64c5c6d19d0b6e15

Source-Link: https://github.com/googleapis/googleapis-gen/commit/9afd8c33d4cae610b75fa4999264ea8c8c66b9d2
Copy-Tag: eyJwIjoiLmdpdGh1Yi8uT3dsQm90LnlhbWwiLCJoIjoiOWFmZDhjMzNkNGNhZTYxMGI3NWZhNDk5OTI2NGVhOGM4YzY2YjlkMiJ9

* 🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

---------

Co-authored-by: Owl Bot <gcf-owl-bot[bot]@users.noreply.github.com>

parthea committed:
fix: remove setup.cfg configuration for creating universal wheels (#1030)

Co-authored-by: Daniel Sanche <sanche@google.com>

daniel-sanche committed:
chore(docs): add BulkWriter to docs (#1033)

* chore(docs): add BulkWriter to docs

* 🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

* 🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

* removed unintentional files

---------

Co-authored-by: Owl Bot <gcf-owl-bot[bot]@users.noreply.github.com>

gcf-owl-bot[bot] committed:
chore(python): fix incorrect import statement in README (#1034)

Source-Link: https://github.com/googleapis/synthtool/commit/87677404f85cee860588ebe2c352d0609f683d5d
Post-Processor: gcr.io/cloud-devrel-public-resources/owlbot-python:latest@sha256:023a21377a2a00008057f99f0118edadc30a19d1636a3fee47189ebec2f3921c

Co-authored-by: Owl Bot <gcf-owl-bot[bot]@users.noreply.github.com>

gcf-owl-bot[bot] committed:
chore(python): remove .gitignore from templates (#1036)

* chore(python): remove .gitignore from templates

Source-Link: https://github.com/googleapis/synthtool/commit/419d94cdddd0d859ac6743ffebd177693c8a027f
Post-Processor: gcr.io/cloud-devrel-public-resources/owlbot-python:latest@sha256:a7aef70df5f13313ddc027409fc8f3151422ec2a57ac8730fce8fa75c060d5bb

* Remove replacement in owlbot.py

---------

Co-authored-by: Owl Bot <gcf-owl-bot[bot]@users.noreply.github.com>
Co-authored-by: Anthonios Partheniou <partheniou@google.com>

gcf-owl-bot[bot] committed:
chore(python): remove CONTRIBUTING.rst from templates (#1038)

* chore(python): remove CONTRIBUTING.rst from templates

Source-Link: https://github.com/googleapis/synthtool/commit/c96fb118e03c2b50d50fe17c1d0845479a0cfa9a
Post-Processor: gcr.io/cloud-devrel-public-resources/owlbot-python:latest@sha256:710b70faff81151657d89db6e028c23a1051787598c8276bdd8eef25c92da8ab

* 🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

* 🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

* Remove replacement in owlbot.py

---------

Co-authored-by: Owl Bot <gcf-owl-bot[bot]@users.noreply.github.com>
Co-authored-by: Anthonios Partheniou <partheniou@google.com>

gcf-owl-bot[bot] committed:
chore(python): remove noxfile.py from templates (#1041)

* chore(python): remove noxfile.py from templates

Source-Link: https://github.com/googleapis/synthtool/commit/776580213a73a04a3ff4fe2ed7f35c7f3d63a882
Post-Processor: gcr.io/cloud-devrel-public-resources/owlbot-python:latest@sha256:25de45b58e52021d3a24a6273964371a97a4efeefe6ad3845a64e697c63b6447

* 🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

* remove replacements in owlbot.py

---------

Co-authored-by: Owl Bot <gcf-owl-bot[bot]@users.noreply.github.com>
Co-authored-by: Anthonios Partheniou <partheniou@google.com>

release-please[bot] committed:
chore(main): release 2.20.2 (#1031)

Co-authored-by: release-please[bot] <55107282+release-please[bot]@users.noreply.github.com>

gkevinzheng committed:
Fixed unit tests

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

gkevinzheng commented:
Closing this PR; #1050 should be the same PR without all the merge conflicts and rebase issues.
